File: apps/comms_web/lib/comms_web/presenters/calls.ex
defmodule CommsWeb.Presenters.Calls do
  @moduledoc false

  alias CommsCore.AudioCalls.{CallSessionView, CallView}

  def audio_call(%CallView{} = call) do
    %{
      id: call.id,
      tenant_id: call.tenant_id,
      conversation_id: call.conversation_id,
      started_by_user_id: call.started_by_user_id,
      ended_by_user_id: call.ended_by_user_id,
      media_kind: call.media_kind,
      status: call.status,
      started_at: call.started_at,
      expires_at: call.expires_at,
      ended_at: call.ended_at,
      end_reason: call.end_reason,
      version: call.version,
      can_end: call.can_end
    }
  end

  def call_session(%CallSessionView{} = call) do
    %{
      id: call.id,
      conversation_id: call.conversation_id,
      started_by_user_id: call.started_by_user_id,
      ended_by_user_id: call.ended_by_user_id,
      media_kind: call.media_kind,
      status: call.status,
      started_at: call.started_at,
      expires_at: call.expires_at,
      ended_at: call.ended_at,
      end_reason: call.end_reason,
      duration_seconds: call.duration_seconds,
      can_end: call.can_end
    }
  end
end

File: clients/web/src/types/calls.ts
export type CallMediaKind = "audio" | "video";

export interface Call {
  id: string;
  conversation_id: string;
  started_by_user_id: string;
  /** Omitted only by legacy audio broadcasts during a rolling deployment. */
  media_kind?: CallMediaKind;
  status: "active" | "ended";
  started_at: string;
  expires_at: string;
  ended_at?: string | null;
  can_end: boolean;
}

export interface CallRealtimeEvent {
  id: string;
  conversation_id: string;
  started_by_user_id: string;
  /** Omitted only by legacy audio broadcasts during a rolling deployment. */
  media_kind?: CallMediaKind;
  status: "active" | "ended";
  started_at: string;
  expires_at: string;
  ended_by_user_id?: string | null;
  ended_at?: string | null;
  end_reason?: string | null;
}

export interface CallIceServer {
  urls: string[];
  username?: string;
  credential?: string;
}

export interface CallCredential {
  server_url: string;
  participant_token: string;
  expires_in: number;
  ice_servers?: CallIceServer[];
}

export interface CallSessionResponse {
  data: Call;
  credential: CallCredential;
}

export interface CallParticipantState {
  id: string;
  user_id: string;
  status: "admitted";
  hand_raised: boolean;
  hand_raised_at: string | null;
}

export type CallsScope = "active" | "recent";

export interface CallSummary {
  id: string;
  conversation_id: string;
  started_by_user_id: string;
  ended_by_user_id: string | null;
  media_kind: CallMediaKind;
  status: "active" | "ending" | "ended";
  started_at: string;
  expires_at: string;
  ended_at: string | null;
  end_reason: string | null;
  duration_seconds: number;
  can_end: boolean;
}

export interface CallsPageResponse {
  data: CallSummary[];
  page: {
    limit: number;
    has_more: boolean;
    next_cursor: string | null;
  };
}

export interface CallsQueryOptions {
  scope?: CallsScope;
  media_kind?: CallMediaKind;
  limit?: number;
  cursor?: string | null;
}

/** Compatibility aliases for consumers migrating from the audio-only surface. */
export type AudioCall = Call;

export type AudioCallRealtimeEvent = CallRealtimeEvent;

export type AudioCallCredential = CallCredential;

export type AudioCallSessionResponse = CallSessionResponse;
